File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
import java.util.regex.*;
import java.util.stream.*;

class Solution {
    public int solution(String dartResult) {
        String reg = "([0-9]+)([SDT])([*#]?)";
        int[] score = new int[3];

        Pattern p = Pattern.compile(reg);
        Matcher matcher = p.matcher(dartResult);
        int idx = 0;
        while (matcher.find()) {
            String group = matcher.group(1);
            // System.out.println("group = " + group);
            int nowScore = Integer.parseInt(matcher.group(1));
            String bonus = matcher.group(2);
            String option = matcher.group(3);

            // 보너스
            if (bonus.equals("S")) {
                score[idx] = nowScore * 1;
            }
            else if (bonus.equals("D")) {
                score[idx] = (int) Math.pow(nowScore,2);
            }
            else {
                score[idx] = (int) Math.pow(nowScore, 3);
            }

            // 옵션
            if (option.equals("*")) {
                if (idx != 0) score[idx-1] = score[idx-1] * 2;
                score[idx] = score[idx] * 2;
            } else if (option.equals("#")) {
                score[idx] = score[idx] * (-1);
            } else {
            }

            idx++;
        }

        // System.out.println(Arrays.toString(score));
        return Arrays.stream(score).sum();
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
import re
def solution(dartResult):
    answer = [0,0,0]
    
    for idx, result in enumerate(re.findall('(\d+)(S|D|T)(\*|\#)?',dartResult)):
        score = int(result[0])
        bonus = result[1]
        option = result[2]
        # print(score,bonus,option)
        score = score if bonus == 'S' else score ** 2 if bonus =='D' else score ** 3
        
        if option == '*':
            score = score * 2
            if idx > 0:
                answer[idx-1] = answer[idx-1] * 2
        elif option == '#':
            score = score * (-1)
        else:
            pass
        
        answer[idx] = score
    
    print(answer)
    
    return sum(answer)